src/sentry/testutils/silo.py

```diff
@@ -231,129 +234,79 @@ def test_something(self):
-    # identifiers are not snowflake ids.
-    restrict_role(role="postgres_unprivileged", model=OrganizationMember, revocation_type="DELETE")
-
-    restrict_role(
-        role="postgres_unprivileged", model=OrganizationMemberMapping, revocation_type="INSERT"
-    )
+        fence_value = _fencing_counters[using]
+        conn.execute("SELECT %s", [f"start_role_override_{fence_value}"])
+        try:
+            yield
+        finally:
+            conn.execute("SELECT %s", [f"end_role_override_{fence_value}"])
 
 
-# Listen to django's migration signal so that we're not trapped inside
-# test method transactions.
-post_migrate.connect(create_model_role_guards, dispatch_uid="create_model_role_guards", weak=False)
+fence_re = re.compile(r"select\s*\'(?P<operation>start|end)_role_override", re.IGNORECASE)
 
 
-def restrict_role(role: str, model: Any, revocation_type: str, using: str = "default") -> None:
-    if router.db_for_write(model) != using:
-        return
+def protected_table(table: str, operation: str) -> re.Pattern:
+    return re.compile(f'{operation}[^"]+"{table}"', re.IGNORECASE)
 
-    using = router.db_for_write(model)
-    with get_connection(using).cursor() as connection:
-        connection.execute(f"REVOKE {revocation_type} ON public.{model._meta.db_table} FROM {role}")
 
+_protected_operations: List[re.Pattern] = []
 
-def protected_table(table: str, operation: str) -> re.Pattern:
-    return re.compile(f'{operation}[^"]+"{table}"', re.IGNORECASE)
 
+def get_protected_operations() -> List[re.Pattern]:
+    if len(_protected_operations):
+        return _protected_operations
 
-protected_operations = (
-    protected_table("sentry_organizationmember", "insert"),
-    protected_table("sentry_organizationmember", "update"),
-    protected_table("sentry_organizationmember", "delete"),
-    protected_table("sentry_organization", "insert"),
-    protected_table("sentry_organization", "update"),
-    protected_table("sentry_organizationmapping", "insert"),
-    protected_table("sentry_organizationmapping", "update"),
-    protected_table("sentry_organizationmembermapping", "insert"),
-)
+    # Protect Foreign Keys using hybrid cloud models from being deleted without using the
+    # privileged user. Deletion should only occur when the developer is actively aware
+    # of the need to generate outboxes.
+    seen_models: MutableSet[type] = set()
+    for app_config in apps.get_app_configs():
+        for model in iter_models(app_config.name):
+            for field in model._meta.fields:
+                if not isinstance(field, HybridCloudForeignKey):
+                    continue
+                fk_model = field.foreign_model
+                if fk_model is None or fk_model in seen_models:
+                    continue
+                seen_models.add(fk_model)
+                _protected_operations.append(protected_table(fk_model._meta.db_table, "delete"))
```

Comment:
I missed adding these guards in the initial pass, but noticed these were missing when removing the role based methods.

Reply:
There should also be a test, somewhere, I wrote that fails if a raw deletion succeeds without throwing an exception.